lib/CL/devices/cuda/pocl-ptx-gen.cc

```diff
@@ -0,0 +1,691 @@
+   of this software and associated documentation files (the "Software"), to deal
+   in the Software without restriction, including without limitation the rights
+   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+#include "config.h"
+
+#include "common.h"
+#include "pocl-ptx-gen.h"
+#include "pocl.h"
+#include "pocl_file_util.h"
+#include "pocl_runtime_config.h"
+
+#include "llvm/Bitcode/BitcodeReader.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/TargetRegistry.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Target/TargetOptions.h"
+#include "llvm/Transforms/IPO.h"
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+#include <set>
+
+namespace llvm {
+extern ModulePass *createNVVMReflectPass(const StringMap<int> &Mapping);
+}
+
+// TODO: Should these be proper passes?
+void pocl_add_kernel_annotations(llvm::Module *module, const char *kernel);
+void pocl_cuda_fix_printf(llvm::Module *module);
+void pocl_cuda_link_libdevice(llvm::Module *module, const char *kernel,
+                              const char *gpu_arch);
+void pocl_gen_local_mem_args(llvm::Module *module, const char *kernel);
+void pocl_insert_ptx_intrinsics(llvm::Module *module);
+void pocl_map_libdevice_calls(llvm::Module *module);
+
+int pocl_ptx_gen(const char *bc_filename, const char *ptx_filename,
+                 const char *kernel_name, const char *gpu_arch) {
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> buffer =
+      llvm::MemoryBuffer::getFile(bc_filename);
+  if (!buffer) {
+    POCL_MSG_ERR("[CUDA] ptx-gen: failed to open bitcode file\n");
+    return 1;
+  }
+
+  // Load bitcode
+  llvm::LLVMContext context;
+  llvm::Expected<std::unique_ptr<llvm::Module>> module =
+      parseBitcodeFile(buffer->get()->getMemBufferRef(), context);
+  if (!module) {
+    POCL_MSG_ERR("[CUDA] ptx-gen: failed to load bitcode\n");
+    return 1;
+  }
+
+  llvm::InitializeAllTargets();
```

Comment:
I'm sure there was a reason I added these, but it seems to be working without them now, so I'll remove them.